File: api/models/points.go
package models

import (
	"slices"
	"time"

	"github.com/sebboness/yektaspoints/util"
)

type PointRequestType string

const PointRequestTypeAdd PointRequestType = "ADD"
const PointRequestTypeSubtract PointRequestType = "SUBTRACT"
const PointRequestTypeCashout PointRequestType = "CASHOUT"

var PointSubtractTypes = []PointRequestType{
	PointRequestTypeSubtract,
	PointRequestTypeCashout,
}

type PointRequestDecision string

const PointRequestDecisionApprove PointRequestDecision = "APPROVE"
const PointRequestDecisionDeny PointRequestDecision = "DENY"

type PointStatus string

const PointStatusWaiting = "WAITING"
const PointStatusSettled = "SETTLED"

var ValidPointRequestDecisions = []PointRequestDecision{
	PointRequestDecisionApprove,
	PointRequestDecisionDeny,
}

type Point struct {
	ID           string       `json:"id" dynamodbav:"id"`
	UserID       string       `json:"user_id" dynamodbav:"user_id"`
	Status       PointStatus  `json:"status" dynamodbav:"status"`
	Points       int32        `json:"points" dynamodbav:"points"`
	Balance      *int32       `json:"balance" dynamodbav:"balance,omitempty"`
	CreatedOnStr string       `json:"-" dynamodbav:"created_on"`
	UpdatedOnStr string       `json:"-" dynamodbav:"updated_on"`
	CreatedOn    time.Time    `json:"created_on" dynamodbav:"-"`
	UpdatedOn    time.Time    `json:"updated_on" dynamodbav:"-"`
	Request      PointRequest `json:"request" dynamodbav:"request"`
}

type PointRequest struct {
	DecidedByUserID string               `json:"decided_by_user_id" dynamodbav:"decided_by_user_id,omitempty"`
	DecidedOnStr    string               `json:"-" dynamodbav:"decided_on,omitempty"`
	DecidedOn       time.Time            `json:"decided_on" dynamodbav:"-"`
	Decision        PointRequestDecision `json:"decision" dynamodbav:"decision,omitempty"`
	ParentNotes     string               `json:"parent_notes" dynamodbav:"parent_notes,omitempty"`
	Reason          string               `json:"reason" dynamodbav:"reason,omitempty"`
	Type            PointRequestType     `json:"type" dynamodbav:"type"`
}

type QueryPointsFilter struct {
	CreatedOn  DateFilter
	UpdatedOn  DateFilter
	Statuses   []PointStatus
	Types      []PointRequestType
	Attributes []string // Which attributes to project in the query
}

type PointSummary struct {
	ID              string               `json:"id"`
	UserID          string               `json:"user_id"`
	ParentNotes     string               `json:"parent_notes"`
	Reason          string               `json:"reason"`
	Points          int32                `json:"points"`
	Type            PointRequestType     `json:"type"`
	UpdatedOn       time.Time            `json:"updated_on"`
	DecidedByUserID string               `json:"decided_by_user_id"`
	Decision        PointRequestDecision `json:"decision" dynamodbav:"decision,omitempty"`
}

type UserPoints struct {
	Balance             int32          `json:"balance"`
	PointsLast7Days     int32          `json:"points_last_7_days"`
	PointsLostLast7Days int32          `json:"points_lost_last_7_days"`
	RecentCashouts      []PointSummary `json:"recent_cashouts"`
	RecentRequests      []PointSummary `json:"recent_requests"`
	RecentPoints        []PointSummary `json:"recent_points"`
}

type PointBalance struct {
	ID      string `json:"id" dynamodbav:"id"`
	UserID  string `json:"user_id" dynamodbav:"user_id"`
	Balance int32  `json:"balance" dynamodbav:"balance"`
}

func (p *Point) ParseTimes() {
	if p.CreatedOnStr != "" {
		p.CreatedOn = util.ParseTime_RFC3339Nano(p.CreatedOnStr)
	}
	if p.UpdatedOnStr != "" {
		p.UpdatedOn = util.ParseTime_RFC3339Nano(p.UpdatedOnStr)
	}
	if p.Request.DecidedOnStr != "" {
		p.Request.DecidedOn = util.ParseTime_RFC3339Nano(p.Request.DecidedOnStr)
	}
}

func (p *Point) ToPointSummary() PointSummary {
	return PointSummary{
		ID:              p.ID,
		UserID:          p.UserID,
		ParentNotes:     p.Request.ParentNotes,
		Points:          p.Points,
		Reason:          p.Request.Reason,
		UpdatedOn:       p.UpdatedOn,
		Type:            p.Request.Type,
		DecidedByUserID: p.Request.DecidedByUserID,
		Decision:        p.Request.Decision,
	}
}

func ToPointSummaries(points []Point) []PointSummary {
	summaries := make([]PointSummary, len(points))
	for idx, p := range points {
		summaries[idx] = p.ToPointSummary()
	}
	return summaries
}

// IsSubtractType returns true if the given point request type is a subtraction type (i.e. SUBTRACT or CASHOUT)
func IsSubtractType(reqType PointRequestType) bool {
	return slices.Contains(PointSubtractTypes, reqType)
}

File: web/lib/models/Points.ts
export enum PointDecision {
    APPROVE = "APPROVE",
    DENY = "DENY",
};

export enum PointStatus {
    SETTLED = "SETTLED",
    WAITING = "WAITING",
};

export enum PointRequestType {
    ADD = "ADD",
    SUBTRACT = "SUBTRACT",
    CASHOUT = "CASHOUT",
};

export type Point = {
    id: string
    user_id: string
    status: string
    points: number
    balance?: number
    created_on: string
    updated_on: string
    request: PointRequest
}

export type PointsList = {
    points: Point[];
}

export interface PointRequest {
    decided_by_user_id?: string
    decided_on?: string
    decision?: string
    parent_notes?: string
    reason?: string
    type: string
}

export type PointSummary = {
    id: string
    parent_notes?: string
    reason?: string
    points: number
    type: string
    updated_on: string
    decided_by_user_id?: string
    decision?: string
};

export type UserPoints = {
    balance: number
    points_last_7_days: number
    points_lost_last_7_days: number
    recent_cashouts: PointSummary[]
    recent_requests: PointSummary[]
    recent_points: PointSummary[]
};

export type RequestPointsRequest = {
    points: number;
    reason: string;
}

export type RequestPointsResponse = {
    point: Point;
    point_summary: PointSummary;
}

export const mapPointToSummary = (p: Point): PointSummary => ({
    id: p.id,
    parent_notes: p.request.parent_notes,
    reason: p.request.reason,
    points: p.points,
    type: p.request.type,
    updated_on: p.updated_on,
    decided_by_user_id: p.request.decided_by_user_id,
    decision: p.request.decision,
});

export const mapPointsToSummaries = (points: Point[]): PointSummary[] => 
    Object.values(points).map((p => mapPointToSummary(p)));
